lib/src/git.rs

```diff
@@ -553,38 +560,67 @@ impl<'a> RemoteCallbacks<'a> {
         // TODO: We should expose the callbacks to the caller instead -- the library
         // crate shouldn't read environment variables.
         callbacks.credentials(move |url, username_from_url, allowed_types| {
+            let span = tracing::debug_span!("RemoteCallbacks.credentials");
+            let _ = span.enter();
+
             let git_config = git2::Config::open_default();
             let credential_helper = git_config
                 .and_then(|conf| git2::Cred::credential_helper(&conf, url, username_from_url));
             if let Ok(creds) = credential_helper {
+                tracing::debug!("using credential_helper");
                 return Ok(creds);
             } else if let Some(username) = username_from_url {
                 if allowed_types.contains(git2::CredentialType::SSH_KEY) {
-                    if std::env::var("SSH_AUTH_SOCK").is_ok()
-                        || std::env::var("SSH_AGENT_PID").is_ok()
+                    if let Some((ssh_auth_sock, ssh_agent_pid)) = std::env::var("SSH_AUTH_SOCK")
+                        .ok()
+                        .zip(std::env::var("SSH_AGENT_PID").ok())
                     {
-                        return git2::Cred::ssh_key_from_agent(username);
+                        tracing::debug!(
+                            username,
+                            ssh_auth_sock,
+                            ssh_agent_pid,
+                            "using ssh_key_from_agent"
+                        );
+                        return git2::Cred::ssh_key_from_agent(username).map_err(|err| {
+                            tracing::error!(err = %err);
+                            err
+                        });
                     }
                     if let Some(ref mut cb) = self.get_ssh_key {
                         if let Some(path) = cb(username) {
-                            return git2::Cred::ssh_key(username, None, &path, None);
+                            tracing::debug!(username, path = ?path, "using ssh_key");
+                            return git2::Cred::ssh_key(username, None, &path, None).map_err(
+                                |err| {
+                                    tracing::error!(err = %err);
+                                    err
+                                },
+                            );
                         }
                     }
                 }
                 if allowed_types.contains(git2::CredentialType::USER_PASS_PLAINTEXT) {
                     if let Some(ref mut cb) = self.get_password {
                         if let Some(pw) = cb(url, username) {
-                            return git2::Cred::userpass_plaintext(username, &pw);
+                            tracing::debug!(username, "using userpass_plaintext");
```

Comment:
Could note here that we're getting the username from the URL.

Reply:
Added!